File: backend/src/services/groupGame.service.js
/**
 * Phase 3.2.4: Group Game Service
 * Handles group game session management, turn-based gameplay, and score aggregation
 */

import GroupActivity from '../models/GroupActivity.js';
import GameScore from '../models/GameScore.js';
import User from '../models/User.js';
import logger from '../config/logger.js';

/**
 * Start a group game session
 * @param {string} classId - Class ID
 * @param {string} gameType - Game type
 * @param {string} startedBy - Teacher user ID
 * @param {Array<string>} participantIds - List of student IDs
 * @returns {Promise<Object>} Group activity
 */
export const startGroupGameSession = async (classId, gameType, startedBy, participantIds = []) => {
  try {
    const activity = await GroupActivity.create({
      activityType: 'game',
      classId,
      startedBy,
      metadata: {
        activityId: null, // Will be set after first game
        activityName: gameType,
        duration: 0
      },
      status: 'active'
    });

    // Add participants if provided
    if (participantIds.length > 0) {
      for (const studentId of participantIds) {
        await activity.addParticipant(studentId);
      }
    }

    logger.info(`Group game session started: ${activity._id} for game ${gameType}`);
    return activity;
  } catch (error) {
    logger.error('Start group game session error:', error);
    throw error;
  }
};

/**
 * Record a turn result in group game
 * @param {string} activityId - Group activity ID
 * @param {string} studentId - Student who played
 * @param {number} score - Score achieved
 * @param {Object} gameData - Additional game data
 * @returns {Promise<Object>} Updated activity
 */
export const recordGroupGameTurn = async (activityId, studentId, score, gameData = {}) => {
  try {
    const activity = await GroupActivity.findById(activityId);
    if (!activity) {
      throw new Error('Group activity not found');
    }

    // Update participant score
    await activity.updateParticipantScore(studentId, score, true);

    // Recalculate group results
    await activity.calculateResults();

    logger.info(`Group game turn recorded: Activity ${activityId}, Student ${studentId}, Score ${score}`);
    return activity;
  } catch (error) {
    logger.error('Record group game turn error:', error);
    throw error;
  }
};

/**
 * Get group game session details
 * @param {string} activityId - Group activity ID
 * @returns {Promise<Object>} Activity with populated data
 */
export const getGroupGameSession = async (activityId) => {
  try {
    const activity = await GroupActivity.findById(activityId)
      .populate('classId', 'name')
      .populate('participants.studentId', 'name email grade')
      .populate('startedBy', 'name email')
      .lean();

    if (!activity) {
      throw new Error('Group activity not found');
    }

    return activity;
  } catch (error) {
    logger.error('Get group game session error:', error);
    throw error;
  }
};

/**
 * Complete group game session
 * @param {string} activityId - Group activity ID
 * @returns {Promise<Object>} Final results
 */
export const completeGroupGameSession = async (activityId) => {
  try {
    const activity = await GroupActivity.findById(activityId);
    if (!activity) {
      throw new Error('Group activity not found');
    }

    // Calculate final results
    await activity.calculateResults();

    // Update status
    activity.status = 'completed';
    await activity.save();

    // Distribute XP to all participants
    const totalXP = activity.results.averageScore || 0;
    const participants = activity.participants;

    for (const participant of participants) {
      if (participant.completed) {
        try {
          const user = await User.findById(participant.studentId);
          if (user) {
            // Add XP to user (simplified - can be enhanced)
            // XP calculation can be more sophisticated
            user.progress = user.progress || {};
            // XP is already tracked in GameScore model
          }
        } catch (userError) {
          logger.warn(`Failed to update XP for user ${participant.studentId}:`, userError);
        }
      }
    }

    logger.info(`Group game session completed: ${activityId}`);
    return activity;
  } catch (error) {
    logger.error('Complete group game session error:', error);
    throw error;
  }
};

/**
 * Get aggregated scores for group game
 * @param {string} activityId - Group activity ID
 * @returns {Promise<Object>} Aggregated scores
 */
export const getGroupGameScores = async (activityId) => {
  try {
    // Get all game scores for this group activity
    const scores = await GameScore.find({ groupActivityId: activityId })
      .populate('userId', 'name email grade')
      .sort({ completedAt: -1 })
      .lean();

    const totalScore = scores.reduce((sum, score) => sum + score.score, 0);
    const totalXP = scores.reduce((sum, score) => sum + (score.xpEarned || 0), 0);
    const averageScore = scores.length > 0 ? totalScore / scores.length : 0;

    return {
      activityId,
      totalTurns: scores.length,
      totalScore,
      averageScore: Math.round(averageScore),
      totalXP,
      scores,
      participants: scores.map(s => ({
        studentId: s.userId?._id || s.userId,
        name: s.userId?.name || 'Unknown',
        score: s.score,
        xpEarned: s.xpEarned || 0,
        completedAt: s.completedAt
      }))
    };
  } catch (error) {
    logger.error('Get group game scores error:', error);
    throw error;
  }
};

File: mobile/lib/features/games/services/group_game_service.dart
/// Phase 3.2.4: Group Game Service
/// Handles group game session management and score aggregation

import '../../../core/services/api_service.dart';
import '../../../core/constants/api_endpoints.dart';

class GroupGameService {
  final ApiService _apiService;

  GroupGameService({ApiService? apiService})
      : _apiService = apiService ?? ApiService();

  /// Start a group game session
  Future<Map<String, dynamic>> startGroupGameSession({
    required String classId,
    required String gameType,
    List<String>? participantIds,
  }) async {
    try {
      final response = await _apiService.post(
        ApiEndpoints.groupActivitiesCreate,
        data: {
          'activityType': 'game',
          'classId': classId,
          'metadata': {
            'activityName': gameType,
            'duration': 0,
          },
          if (participantIds != null && participantIds.isNotEmpty)
            'participantIds': participantIds,
        },
      );

      final data = response.data as Map<String, dynamic>;
      if (response.statusCode == 201 && data['success'] == true) {
        return data['data'] as Map<String, dynamic>;
      }
      throw Exception(data['message'] ?? 'Failed to start group game session');
    } catch (e) {
      print('Error starting group game session: $e');
      rethrow;
    }
  }

  /// Record a turn result in group game
  Future<Map<String, dynamic>> recordGroupGameTurn({
    required String activityId,
    required String studentId,
    required int score,
    Map<String, dynamic>? gameData,
  }) async {
    try {
      final response = await _apiService.post(
        ApiEndpoints.groupActivitySubmit(activityId),
        data: {
          'studentId': studentId,
          'score': score,
          'completed': true,
          if (gameData != null) 'gameData': gameData,
        },
      );

      final data = response.data as Map<String, dynamic>;
      if (response.statusCode == 200 && data['success'] == true) {
        return data['data'] as Map<String, dynamic>;
      }
      throw Exception(data['message'] ?? 'Failed to record turn');
    } catch (e) {
      print('Error recording group game turn: $e');
      rethrow;
    }
  }

  /// Get group game session details
  Future<Map<String, dynamic>> getGroupGameSession(String activityId) async {
    try {
      final response = await _apiService.get(
        ApiEndpoints.groupActivityResults(activityId),
      );

      final data = response.data as Map<String, dynamic>;
      if (response.statusCode == 200 && data['success'] == true) {
        return data['data'] as Map<String, dynamic>;
      }
      throw Exception(data['message'] ?? 'Failed to get group game session');
    } catch (e) {
      print('Error getting group game session: $e');
      rethrow;
    }
  }

  /// Complete group game session
  Future<Map<String, dynamic>> completeGroupGameSession(String activityId) async {
    try {
      // Get final results
      final session = await getGroupGameSession(activityId);
      
      // Update status to completed (this would typically be done via backend endpoint)
      // For now, we'll just return the session data
      return session;
    } catch (e) {
      print('Error completing group game session: $e');
      rethrow;
    }
  }

  /// Get aggregated scores for group game
  Future<Map<String, dynamic>> getGroupGameScores(String activityId) async {
    try {
      final session = await getGroupGameSession(activityId);
      
      // Calculate aggregated scores from session data
      final participants = session['participants'] as List<dynamic>? ?? [];
      final totalScore = participants.fold<int>(
        0,
        (sum, p) => sum + ((p['score'] as num?)?.toInt() ?? 0),
      );
      final averageScore = participants.isNotEmpty
          ? (totalScore / participants.length).round()
          : 0;
      final completedCount = participants
          .where((p) => p['completed'] == true)
          .length;

      return {
        'activityId': activityId,
        'totalTurns': participants.length,
        'totalScore': totalScore,
        'averageScore': averageScore,
        'completedCount': completedCount,
        'completionRate': participants.isNotEmpty
            ? ((completedCount / participants.length) * 100).round()
            : 0,
        'participants': participants,
      };
    } catch (e) {
      print('Error getting group game scores: $e');
      rethrow;
    }
  }
}

